src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -22,15 +22,52 @@ internal partial class Binder
             BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
         {
             // PROTOTYPE(tuple-equality) Block in expression tree
-
             TupleBinaryOperatorInfo.Multiple operators = BindTupleBinaryOperatorNestedInfo(node, kind, left, right, diagnostics);
 
-            // PROTOTYPE(tuple-equality) We'll save the converted nodes separately, for the semantic model
-            //BoundExpression convertedLeft = GenerateConversionForAssignment(operators.LeftConvertedType, left, diagnostics);
-            //BoundExpression convertedRight = GenerateConversionForAssignment(operators.RightConvertedType, right, diagnostics);
+            DiagnosticBag discardDiagnostics = DiagnosticBag.GetInstance();
+
+            // The converted types are only used for the semantic model, so we don't need the conversion diagnostics
+            BoundExpression convertedLeft = ApplyConvertedTypes(left, operators, isRight: false, discardDiagnostics);
+            BoundExpression convertedRight = ApplyConvertedTypes(right, operators, isRight: true, discardDiagnostics);
+
+            discardDiagnostics.Free();
 
             TypeSymbol resultType = GetSpecialType(SpecialType.System_Boolean, diagnostics, node);
-            return new BoundTupleBinaryOperator(node, left, right, kind, operators, resultType);
+
+            return new BoundTupleBinaryOperator(node, left, right, convertedLeft, convertedRight, kind, operators, resultType);
+        }
+
+        BoundExpression ApplyConvertedTypes(BoundExpression expr, TupleBinaryOperatorInfo @operator, bool isRight, DiagnosticBag diagnostics)
+        {
+            TypeSymbol convertedType = isRight ? @operator.RightConvertedTypeOpt : @operator.LeftConvertedTypeOpt;
+
+            if (@operator.InfoKind == TupleBinaryOperatorInfo.KindEnum.Multiple && expr.Kind == BoundKind.TupleLiteral)
+            {
+                var multiple = (TupleBinaryOperatorInfo.Multiple)@operator;
+                var tuple = (BoundTupleLiteral)expr;
+                ImmutableArray<BoundExpression> arguments = tuple.Arguments;
+                int length = arguments.Length;
+
+                Debug.Assert(length == multiple.Operators.Length);
+
+                var builder = ArrayBuilder<BoundExpression>.GetInstance(length);
+                for (int i = 0; i < length; i++)
+                {
+                    builder.Add(ApplyConvertedTypes(arguments[i], multiple.Operators[i], isRight, diagnostics));
```

Comment:
>builder.Add(ApplyConvertedTypes(arguments[i], multiple.Operators[i], isRight, diagnostics)); [](start = 20, length = 92)

It feels like, instead of applying element-wise conversions explicitly, we should simply apply conversion to the entire literal when the convertedType is not null. Otherwise, we are actually changing the element types but leaving tuple's tuple's type unchanged, then we are trying to convert that hacked literal that contains possibly inconsistent information. That might or might not give us result that we want. Why take chances? I think that explicitly applying element-wise conversions makes sense only for cases when we don't have convertedType.